### PR TITLE
HPCC-24229 Download WU result as files using streaming

### PR DESCRIPTION
**Pull request by wangkx**
1. Download the result chunk by chunk based on a pre-configured
buffer size limit;
2. When downloading the result in XML, CSV, or JSON format,
call CHttpMessage::sendChunk() for every chunk;
3. When downloading the result in ZIP or GZIP format, the
result is streamed into a temporary file first. Then, the
file is converted to the ZIP or GZIP format and sent out
using IFileIOStream.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review

File: esp/services/ws_workunits/ws_wuresult.cpp [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+    if (outIOS)
+    {
+        outIOS->write(current.length(), current.str());
+        current.clear();
+        return;
+    }
+
+    s.append(current);
+    current.clear();
+    if (s.length() < (closing ? 1 : wuResultDownloadFlushThreshold))
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_wuresult.cpp:51]: Minor, but might be a slightly clearer if this was broken into two lines.  
unsigned threshold = closing ? 1 : wuResultDownloadFlushThreshold;

File: esp/services/ws_workunits/ws_wuresult.cpp [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+            response->setContentType("application/zip");
+        response->send();
+        recursiveRemoveDirectory(workingFolder);
+    }
+#endif
+
+    return true;
+}
+
+bool CWsWuResultOutHelper::canStreaming()
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_wuresult.cpp:144]: Better name might be "canSupportStreaming()" or "canStream()".

File: esp/services/ws_workunits/ws_wuresult.cpp [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+    SCMStringBuffer logicalNameBuf;
+    result->getResultLogicalName(logicalNameBuf);
+
+    Owned<INewResultSet> rs;
+    Owned<IResultSetFactory> resultSetFactory = getSecResultSetFactory(context->querySecManager(),
+        context->queryUser(), context->queryUserId(), context->queryPassword());
+    if (logicalNameBuf.length())
+    {
+        rs.setown(resultSetFactory->createNewFileResultSet(logicalNameBuf.str(), cw->queryClusterName())); //MORE is this wrong cluster?
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_wuresult.cpp:347]: The comment is at least 6 years old... would we know by now if it was the wrong cluster?

Comment (wangkx) [reply esp/services/ws_workunits/ws_wuresult.cpp:347]: We still do not know. But, we try it because there is no other option.

**Review by afishbeck — APPROVED** [review]
@wangkx looks fine

**Pull request merged**